### PR TITLE
Replace `request` with `got`

### PR DESCRIPTION
**Pull request by hankewyczz**
image

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -9,6 +9,7 @@ jobs:
   tests:
     name: Tests
     runs-on: ubuntu-latest
+    timeout-minutes: 10
```
Comment (hankewyczz) [inline .github/workflows/test.yaml:12]: This, and `detectOpenHandles`, were both added to handle the case that something doesn't properly close out. 

During this PR, I had issues where a dependency wouldn't kill itself, so Jest would stay open until Github forcibly killed it. 

This shouldn't be necessary anymore, but it's useful to keep just in case.

File: package.json [hunk 2]
```diff
@@ -81,18 +81,19 @@
     "atob": "^2.0.3",
     "axios": "^0.25.0",
     "babel-core": "^7.0.0-bridge.0",
+    "cacheable-lookup": "^6.1.0",
```
Comment (hankewyczz) [inline package.json:84]: A replacement for `dnscache` that isn't deprecated, and actually works really well with `got`

File: package.json [hunk 3]
```diff
@@ -81,18 +81,19 @@
...
     "core-js": "^3.0.1",
     "cors": "^2.8.5",
     "cross-env": "^7.0.3",
-    "dnscache": "^1.0.1",
     "domutils": "^2.8.0",
     "dotenv": "^16.0.0",
     "eslint-plugin-deprecation": "^1.2.1",
     "express": "^4.17.3",
     "fs-extra": "^10.0.0",
+    "got": "^11.8.2",
```
Comment (hankewyczz) [inline package.json:96]: Our new requests library!!!

File: package.json [hunk 4]
```diff
@@ -111,6 +112,7 @@
     "request": "^2.81.0",
     "request-promise-native": "^1.0.9",
     "rollbar": "^2.19.4",
+    "tough-cookie": "^4.1.2",
```
Comment (hankewyczz) [inline package.json:115]: Used for storing cookies. Some URLs on the Banner site need specific cookies in order to work.

It goes something like this:
- you open the Banner site, select a specific term, and press "Submit"
- Banner saves a cookie essentially saying "this guy can search for classes in this term"

We want to save that cookie so that we can actually execute those searches. Our old requests library had this built in; the new one doesn't.

File: package.json [hunk 5]
```diff
@@ -170,6 +172,7 @@
     "husky": "^4.3.8",
     "jest": "^27.0.0",
     "jest-environment-node": "^27.0.0",
+    "nock": "^13.3.0",
```
Comment (hankewyczz) [inline package.json:175]: This is used to mock requests while testing - this way, we don't have to rely on an Internet connection.

File: scrapers/classes/parsersxe/tests/classParser.test.js [hunk 6]
```diff
@@ -17,10 +32,15 @@ beforeAll(() => {
   jest
     .spyOn(ClassParser, "getAttributes")
     .mockReturnValue(["innovation", "bizz"]);
+  jest.spyOn(ClassParser, "getFees").mockReturnValue({
```
Comment (hankewyczz) [inline scrapers/classes/parsersxe/tests/classParser.test.js:35]: Before this PR, requests to this method were making real-life requests — something we want to avoid.

File: scrapers/classes/parsersxe/util.ts [hunk 7]
```diff
@@ -94,25 +95,27 @@ async function getCookiesForSearch(termId: string): Promise<req.CookieJar> {
       form: {
         term: termId,
       },
-      cache: false,
+      cacheRequests: false,
+      cookieJar: cookieJar,
```
Comment (hankewyczz) [inline scrapers/classes/parsersxe/util.ts:99]: we now have to manually handle the cookie jar - see my comment in `package.json` for more details

File: scrapers/main.ts [hunk 8]
```diff
@@ -39,7 +40,13 @@ class Main {
       currentTermInfos: currentTermInfos,
     });
 
-    macros.log("Done scraping\n\n".green.underline);
+    const totalTime = Date.now() - start;
```
Comment (hankewyczz) [inline scrapers/main.ts:43]: This has nothing to (directly) do with the requests - I was just using this to measure the time difference between requests before and after my changes. 

I'm leaving this in because it's a feature that's nice to have.

File: scrapers/tests/request.test.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+/*
+ * This file is part of Search NEU and licensed under AGPL3.
```
Comment (hankewyczz) [inline scrapers/tests/request.test.ts:2]: File was renamed from .js

The only major changes are the use of `httpbin.org` for our requests instead of using mocks. We do want to try an actual, live request at least once.

File: services/notifyer.ts [hunk 10]
```diff
@@ -47,7 +47,8 @@ export async function sendNotifications(
     const courseNotifPromises: Promise<void>[] = notificationInfo.updatedCourses
       .map((course) => {
         const courseMessage = generateCourseMessage(course);
-        return courseHashToUsers[course.courseHash].map((user) => {
+        const users = courseHashToUsers[course.courseHash] ?? [];
```
Comment (hankewyczz) [inline services/notifyer.ts:50]: Handles the edge case where `courseHashToUsers` doesn't have any key matching `course.courseHash`. 

This isn't related to requests, just something I saw and fixed.

File: tests/general/utils/macros.test.ts [hunk 11]
```diff
@@ -334,7 +334,7 @@ describe("console statements", () => {
       macros.error("test");
 
       // Called twice per macros.error call
-      expect(console.error).toHaveBeenCalledTimes(12);
+      expect(console.error).toHaveBeenCalledTimes(6);
```
Comment (hankewyczz) [inline tests/general/utils/macros.test.ts:337]: We no longer use `console.trace` in our macros calls, so the number of calls has been halved. The way `console.trace` is used in Node is now different, and it provides us with nothing useful.

File: scrapers/tests/__mocks__/request-promise-native.js [hunk 12]
```diff
@@ -1,67 +0,0 @@
-/*
```
Comment (hankewyczz) [inline scrapers/tests/__mocks__/request-promise-native.js:1]: This was used to mock the requests library for unit testing. 

We now use `nock` for that purpose, so this is no longer useful.

File: scrapers/request.ts [hunk 13]
```diff
@@ -353,7 +318,8 @@ class Request {
     this.openRequests++;
 
     try {
-      return await request(output);
+      const { default: got } = await import("got");
```
Comment (hankewyczz) [inline scrapers/request.ts:321]: Add comment about ESM

**Review by pranavphadke1 — APPROVED** [review]

**Review by Lucas-Dunker — APPROVED** [review]
Looks great! For general knowledge transfer purposes, I'll shoot you a DM on slack so we can go through the PR together sometime if that works for you

File: scrapers/classes/parsersxe/bannerv9Parser.ts [hunk 14]
```diff
@@ -19,7 +19,7 @@ import { ParsedCourseSR, ParsedTermSR } from "../../../types/scraperTypes";
 import { MultiProgressBars } from "multi-progress-bars";
 
 // Only used to query the term IDs, so we never want to use a cached version
-const request = new Request("bannerv9Parser", { cache: false });
+const request = new Request("bannerv9Parser", { cacheRequests: false });
```
Comment (Lucas-Dunker) [inline scrapers/classes/parsersxe/bannerv9Parser.ts:22]: Is `cacheRequests` just the term that `got` uses rather than `cache`?

File: scrapers/classes/parsersxe/termParser.ts [hunk 15]
```diff
@@ -185,24 +186,25 @@ class TermParser {
...
-    const cookiejar = await util.getCookiesForSearch(termId);
+    const cookieJar = await util.getCookiesForSearch(termId);
     // second, get the total number of sections in this semester
     try {
       return (await this.concatPagination(async (offset, pageSize) => {
         const req = await request.get(
           "https://nubanner.neu.edu/StudentRegistrationSsb/ssb/searchResults/searchResults",
           {
-            qs: {
+            searchParams: {
```
Comment (Lucas-Dunker) [inline scrapers/classes/parsersxe/termParser.ts:196]: I assumed that this change is just due to the way that `got` names its params

File: package.json (hunk 2 above)
Comment (hankewyczz) [reply package.json:74]: `express` is a web framework, not really a requests library. 

Between requests libraries — [this comparison sums it up](https://github.com/sindresorhus/got#comparison). Obviously, that's a biased source (since it's from `got`), but it's not wrong. Essentially:
- `got` isn't a **bad** choice. It's in active development and very widely used. 
- It supports everything we need, and some things we might want to include in the future (eg. RFC-compliant caching)
- It's pretty simple to use

We could've gone with something else with no problem, but `got` just ended up being the choice.

File: scrapers/classes/parsersxe/bannerv9Parser.ts (hunk 14 above)
Comment (hankewyczz) [reply scrapers/classes/parsersxe/bannerv9Parser.ts:22]: No — `cache` and `cacheRequests` are both our own options. However, `got` has its own `cache` option, and I didn't want people to get confused about this - is it `got`'s `cache` or our own?

File: scrapers/classes/parsersxe/subjectAbbreviationParser.ts [hunk 16]
```diff
@@ -33,14 +33,12 @@ async function requestSubjects(termId: string): Promise<SubjectDescription[]> {
...
-  const response = await request.get(subjectUrl, {
-    json: true,
-  });
+  const response = await request.get(subjectUrl);
 
   if (response.statusCode !== 200) {
     macros.error(`Problem with request for subjects ${subjectUrl}`);
   }
-  return response.body;
+  return JSON.parse(response.body);
```
[the remark this answers is not in the document]
Comment (hankewyczz) [reply scrapers/classes/parsersxe/subjectAbbreviationParser.ts:41]: Sure! Basically, our old library parsed the body to JSON if we specified `json: true`. This library _can_ do the same, but:

- Our requests library always returned the body of the request. We don't want to change this behavior (at least not in the PR) because it'd be a large breaking change
- In `got`, the `json()` method can only be called on a `Response` object. This object contains everything - we don't want to return that.

So the issue becomes - how can we convert to json without exposing the `Response` object and breaking compatibility?

This was my approach. In theory - we could add a `json` option, and then inside the request function, call `JSON.parse` on the response body if that option is true. That might be better - I'm open to changes. My initial thinking was - let's make this simple. A body is always a string, and after that, you do what you want with it.

File: scrapers/classes/parsersxe/termParser.ts (hunk 15 above)
Comment (hankewyczz) [reply scrapers/classes/parsersxe/termParser.ts:196]: I vividly remember writing a comment SOMEWHERE about this, but I can't find it. Oh well. That's exactly it. Read more [here](https://github.com/sindresorhus/got/blob/e24b89669931b36530219b9f49965d07da25a7e6/documentation/migration-guides/request.md#renamed-options)

File: scrapers/classes/parsersxe/tests/bannerv9Parser.test.ts [hunk 17]
```diff
@@ -7,6 +7,29 @@ import prisma from "../../../../services/prisma";
...
+    },
+    {
+      code: "2",
+      description: "Summer 2022 Law Semester",
+    },
+  ])
+  .persist();
+
+afterAll(() => {
+  scope.persist(false);
```
[the remark this answers is not in the document]
Comment (hankewyczz) [reply scrapers/classes/parsersxe/tests/bannerv9Parser.test.ts:31]: > When you setup an interceptor for a URL and that interceptor is used, it is removed from the interceptor list. This means that you can intercept 2 or more calls to the same URL and return different things on each of them. It also means that you must setup one interceptor for each request you are going to have, otherwise nock will throw an error because that URL was not present in the interceptor list. If you don’t want interceptors to be removed as they are used, you can use the [.persist()](https://github.com/nock/nock#persist) method.

That's why we set `persist()` at first - we want `nock` to use this interceptor for every matching request we have in the file.

However - we don't want this affecting other files! So at the end:
> If you want to stop persisting an individual persisted mock you can call persist(false):

File: scrapers/classes/parsersxe/tests/subjectAbbreviationParser.test.js [hunk 18]
```diff
@@ -1,5 +1,24 @@
...
+    {
+      code: "AVM",
+      description: "Adv Manufacturing System - CPS",
+    },
+  ])
+  .persist();
+
+afterAll(() => {
+  scope.persist(false);
+});
```
[the remark this answers is not in the document]
Comment (hankewyczz) [reply scrapers/classes/parsersxe/tests/subjectAbbreviationParser.test.js:21]: See above comment - does that answer it?

File: scrapers/classes/parsersxe/bannerv9Parser.ts (hunk 14 above)
Comment (sebwittr) [reply scrapers/classes/parsersxe/bannerv9Parser.ts:22]: Is there any place where cacheRequests is being passed in as true?

File: scrapers/classes/parsersxe/subjectAbbreviationParser.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (sebwittr) [reply scrapers/classes/parsersxe/subjectAbbreviationParser.ts:41]: I think either option is fine. Personally I think it would be easier to have a json option.

**Review by sebwittr — APPROVED** [review]
Everything looks really good! The comments helped a lot with understanding this PR, nice job!

File: scrapers/classes/parsersxe/bannerv9Parser.ts (hunk 14 above)
Comment (sebwittr) [reply scrapers/classes/parsersxe/bannerv9Parser.ts:22]: Disregard that, im being dumb^

File: scrapers/classes/parsersxe/subjectAbbreviationParser.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (hankewyczz) [reply scrapers/classes/parsersxe/subjectAbbreviationParser.ts:41]: Sure! Like I said — that's an open issue, could be discussed. Either way, this isn't a change that should be done in this PR, because that'll make this already too-complicated PR insanely overcomplicated
